[MLMC] New application structure

riccardotosi opened the pull request:
This PR's aim is to redefine the structure of the application. The structure is the following:

* python_scripts: contains the utilities to perform the Monte Carlo and Multilevel Monte Carlo algorithms. The test_*.py files are needed to run the application tests without COMPSs,
* tests: contains the tests of the application. In a successive PR tests for specific function of the application (such as the computation of central moments) will be considered. It contains folders with the corresponding application-test, in this case we have the "poisson_square_2d" folder.
    * poisson_square_2d: this folder contains the analysis stage of the problem, which is derived from the specific application we are considering, and a main, that is the file to run in order to execute the simulation,
        * problem_settings: contains model part and project parameters of each application, the settings of the monte carlo simulation and of the refinement in case of multilevel monte carlo simulation.

The analysis stage `SimulationScenario` is what the user should modify in order to solve a different problem, while all the utilities inside python_scripts and the main file should remain the same.
The application works also using PyCOMPSs.

In the Examples repository we will add in future specific problems (such as compressible potential flow) that are not contained inside the tests folder.

The flag `store_lower_levels_samples` is added, and gives the possibility to the user to decide if storing or not the samples computed through the adaptive refinement.

Commit:
Add new mc_utilities and montecarlo_analysis using new analysis stage

Commit:
Relative parameters path

Commit:
Temporarily add mockups for the new analysis stage

riccardotosi committed:
Changes of TODOs after Riccardo's meeting

Commit:
Merge branch 'MLMC/reduce-executionMLMC-task' into MLMC/new-analysis-stage

Commit:
Merge branch 'MLMC/new-analysis-stage' of https://github.com/KratosMultiphysics/Kratos into MLMC/new-analysis-stage

Commit:
Removing mockups from branch

Commit:
Rename Initialize list function

Commit:
Construct mc_utilities new analysis working in python

Commit:
Move to poisson analysis stage the poisson eq

Commit:
Creation generator file

Commit:
Add a TODO

Commit:
Rename poisson analysis

Commit:
Add boolean serialization model and parameters

Commit:
Merge branch 'master' into MLMC/new-analysis-stage

Commit:
Rename meah sizes

Commit:
Prepare to cmlmc

Commit:
Creation of poisson analysis stage,MLMC and MC main, adaptation cmlmc utilities to new analysis stage. Working with pycompss

Commit:
Set relative path

Commit:
import simulation_definition

Commit:
move files to test_examples flder

Commit:
Creation test_examples

Commit:
Set relative path

Commit:
Move runscreening() inside run()

Commit:
Update cmlmc and mc utilities to the new analysis stage

Commit:
Remove montecarlo and multilevel montecarlo analysis: now they are simulation_definition.py

Commit:
import updated cmlmc and mc utilities

Commit:
Update importing analysis stage

riccardotosi committed:
Rename generator file

riccardotosi committed:
Update to new generator name

Commit:
Update to new Kratos python import module

Commit:
Change name to UpperCamelCase

riccardotosi committed:
Merge branch 'master' into MLMC/new-analysis-stage

riccardotosi committed:
Use double ' instead of single '

riccardotosi committed:
Use # when needed and define inputs/outputs of the functiongit status

riccardotosi committed:
Use # when needed and define inputs/outputs of the function

riccardotosi committed:
Add TODOs

riccardotosi committed:
Updating input/output of functions

riccardotosi committed:
Update input/output description

riccardotosi committed:
Update function name to ComputeSampleCentralMomentsFromScratch()

riccardotosi committed:
Use StatisticalVariable.number_samples[level] instead of MonteCarlo.StatisticalVariable.number_samples[level]

riccardotosi committed:
Remove two TODOs done

riccardotosi committed:
Rename moments from scratch

riccardotosi committed:
Removal of TODO from mc and cmlmc and insert in aux_variables utilities

riccardotosi committed:
Renaming file to statistical_variabl_utilities

riccardotosi committed:
Starting to use # instead of ''

riccardotosi committed:
Finishing modifying input/output mc utilities

riccardotosi committed:
Update check convergence variable names

Commit:
Add check ComputeTolerancei TODO

Commit:
Merge branch 'master' into MLMC/new-analysis-stage

riccardotosi committed:
Merge branch 'master' into MLMC/new-analysis-stage

riccardotosi committed:
Use # instead of '

riccardotosi committed:
Change cmlmc symbol from mlmc to cmlmc

riccardotosi committed:
Retrieve model part name from project parameters and not writing it

Commit:
Merge branch 'MLMC/new-analysis-stage' of https://github.com/KratosMultiphysics/Kratos into MLMC/new-analysis-stage

Commit:
Modify project parameters name

Commit:
Merge branch 'master' into MLMC/new-analysis-stage

Commit:
Renaming model parts

Commit:
Uniform python style

Commit:
Remove unused duplicated parameters

Commit:
Modify to use specific class defined in specific path

Commit:
Starting to modify tests

Commit:
Updating tests for new analysis stage format

Commit:
Renaming gid output

Commit:
Updating remove files but not working

Commit:
Delete files after test

Commit:
Add computation first central moment

Commit:
Rename to central_moment_p

Commit:
Compute tolerance iteration i as in [PNL17]

Commit:
Remove TODOs from test

riccardotosi committed:
Use h-statistics and not sample variance in CheckConvergence function

riccardotosi committed:
Add a TODO

riccardotosi committed:
Add number of samples for current level in ComputeSampleCentralMomentsFromScratch since it cannot be retrieved using COMPSs and modify inputs and outputs of UpdateOnePassMomentsVariance to work with COMPSs

Commit:
Remove TODO: check equivalence central moments from scratch and on the fly

philbucher commented on kratos/python_scripts/KratosUnittest.py:
```diff
@@ -160,10 +160,13 @@ class WorkFolderScope:
 
     def __enter__(self):
         os.chdir(self.scope)
+        sys.path.append(self.scope)
```
changes to the core (esp such a class) affect everyone, please make a separate PR explaining why the changes are needed

btw I am not sure if this can go in, we currently try to remove adding things to the path bcs it messes things up
=> to be discussed in a separate PR

riccardotosi replied:
I discussed with @roigcarlo and he proposed to put this change in this specific PR since it is needed in this specific PR.

roigcarlo replied:
@philbucher As @riccardotosi says this is somehow hard to understand without an specific example which needs to use this functionality. Hance why is in here. but if you preffer we candiscuss it in a separate PR.

But as the topic has already appeared, a little explanation is in order:
The problem here is caused because the specific example we need to run assumes that, apart from the regular MainKratos.py, the directory from which the problem will run has an aditional *.py file. Unfortunatelly right now there is no other way to proced with this. 

Normally this would not be problem, but since the Unittest library runs the tests from an arbitrary palce we need to tell python where to look for this file. Since we already have a function which manipulates the dir it seemed logic to put it here.

philbucher replied:
no worries @roigcarlo :)

Why don't you manually add the path of the test-file to the python-path in the respective test? This should work if the "other" py-file is in the same directory

I am just afraid if we add it by default then we could hide some import-problems :/

=> if my first proposal doesn't work then could you make the adding to the path an optional thing? I.e. a constructor argument defaulting to `False`?

roigcarlo replied:
I like more the second option, maybe we can make a separate context manager just for paths as we have for dirs, I didn't think on that one honestly

philbucher replied:
@roigcarlo up to you :) 
I think here a default arg could make more sense, since the rest of the contextmanager is the same (as far as I understood) 
=> less chances of the implementations to diverge in the future

@riccardotosi I will remove my block as soon as someone reviewed the rest, since it is currently not possible to dismiss reviews and I cannot approve this

Commit:
Use _ instead of instance after Codacy review

Commit:
Remove unused import adaptive refinement (called inside cmlmc utilities)

Commit:
Moving poisson problem inside tests folder

roigcarlo committed:
Making change in the pythonpath optional

roigcarlo committed:
Merge branch 'master' into MLMC/new-analysis-stage

roigcarlo committed:
Fix using unexisting var

roigcarlo committed:
Merge branch 'MLMC/new-analysis-stage' of https://github.com/KratosMultiphysics/Kratos into MLMC/new-analysis-stage

riccardotosi commented:
Thanks @roigcarlo!

riccardotosi committed:
Renaming PyCOMPSs import

riccardotosi committed:
Renaming some non-intuitive variables name

Commit:
Add cphi derived from confidence

Commit:
Correct bug cphi_confidence from < to ==

Commit:
Add 3rd convergence criteria

Commit:
Add error probability that is needed in two convergence criteria (earlier I was using confidence = 1 - error_probability)

Commit:
Do not set error_probability in default MC but evaluate from confidence

Commit:
In total_error_stopping_rule use confidence to compute cphi, not error_probability

riccardotosi committed:
Renaming mean and sample variance

riccardotosi committed:
Update number samples in onepassupdatepowersumsaux_task for MC

riccardotosi committed:
Use only power sums to compute unbiased central moments

philbucher commented:
just to say that for me the changes in the Core are ok
I will release my block as soon as someone else approves this PR, since I cannot approve it

Commit:
Remove S3 absolute

riccardotosi commented:
@RiccardoRossi ping

Commit:
Random input in list

riccardotosi committed:
First raw moment (mean) computed in h statistics 1

Commit:
Add possibility to set max size, default is False

Commit:
Merge branch 'MLMC/new-analysis-stage' of https://github.com/KratosMultiphysics/Kratos into MLMC/new-analysis-stage

Commit:
Add json for xmc and refinement

philbucher commented on applications/MultilevelMonteCarloApplication/python_scripts/mc_utilities.py:
```diff
@@ -5,249 +5,473 @@
 
 # Import the StatisticalVariable class
-from auxiliary_classes_utilities import StatisticalVariable
+from statistical_variable_utilities import StatisticalVariable
```
side note: this way of importing python-scripts is deprecated and will stop working at some point (not soon though) therefore we encourage devs to slowly adapt the new way of importing:
```suggestion
from KratosMultiphysics.MultilevelMonteCarloApplication.statistical_variable_utilities import StatisticalVariable
```

riccardotosi replied:
Thanks

philbucher replied:
You can check the CompressiblePotentialApp, they already use the [new import system](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CompressiblePotentialFlowApplication/CompressiblePotentialFlowApplication.py#L11)

riccardotosi replied:
Ty, done!

philbucher committed:
dummy commit to remove approval

Commit:
Updating import statements

Commit:
New python standards

philbucher commented on applications/MultilevelMonteCarloApplication/MultilevelMonteCarloApplication.py:
```diff
@@ -2,13 +2,10 @@
 from __future__ import print_function, absolute_import, division
 
 # Application dependent names and paths
+from KratosMultiphysics import *
```
```suggestion
import KratosMultiphysics as KM
```

philbucher commented on applications/MultilevelMonteCarloApplication/MultilevelMonteCarloApplication.py:
```diff
@@ -2,13 +2,10 @@
-import inspect
-caller = inspect.stack()[1] # Information about the file that imported this, to check for unexpected imports
-application_importer.ImportApplication(application,application_name,application_folder,caller,  __path__)
+KratosMultiphysics._ImportApplicationAsModule(application, application_name, application_folder, __path__)
```
```suggestion
KM._ImportApplicationAsModule(application, application_name, application_folder, __path__)
```

riccardotosi replied:
Done

philbucher reviewed (commented):
Nice!

If imports of some python-scripts stop working, just add `KratosMultiphysics.MultilevelMonteCarloApplication.` before the name of the python-script, then it should work again

pablobecker commented:
In principle, the @KratosMultiphysics/implementation-committee has no problems with yourchanges to the core. The rest looks opk too. However, we must ask you to separate this single change to the core and put it in its own PR so we can more cleanly evaluate it.

philbucher commented:
@riccardotosi I recommend to take a look at codacy
Some of the suggestions are relevant, esp the ones abt unused imports

roigcarlo commented:
Just passing by to tell that probably is not a good idea to blindly accept codacy suggestions about unused imports in this particular scenario. I can't tell you with a 100% confidence all of the reports are wrong but I suspect some of them might be.

I explain myself. The python code of this PR is not meant to be executed by a "regular" python runtime, but rather by a swarm of them. The problem here appears while constructing the communication of data between instances which is done by pickle and unpickle the python classes. 
The problem is that pickle does not provide a mechanism to store the class "prototype" (let me use js jargon here as I don't know the python equivalent).

With a little example, imagine these scripts:
```python
# Myexecutionbody.py
import pycompps

@pycompss_task
def distributed_task(pickled_object):
     myclass_instance = unpickle(pickled_object)
     print(myclass_instance.id)
```

```python
 # Main.py (pseudocode)
import myclass
import myexecutionbody

myclass_pickled = pickle(myclass("NewInstance"))

myexecutionbody.distributed_task(myclass_pickled)
```

Without entering in the details of the implementation of the pickled class, you can see that if I try to execute the `distributed_task` without having been called by the `Main.py` it is impossible to restore the object as I don't know what it is. That situation should be impossible if running python from your commandline but its its possible with pyCompss.

It happens something similar with the kratos file serialization. If I serialize something into a file which contains classes from the StructuralMechanics, and then I try to load it in another script which only loads the FluidDynamics its going to error.

Commit:
Change import way

Commit:
Delete a print of the settings

roigcarlo committed:
Reverting changes in KratosUnittest

This will break the PR and its intentional

roigcarlo committed:
Merge branch 'MLMC/new-analysis-stage' of https://github.com/KratosMultiphysics/Kratos into MLMC/new-analysis-stage

roigcarlo committed:
Extending WorkFolderScope context manager

Commit:
Merge branch 'core/extending-workfolderscope' into MLMC/new-analysis-stage

roigcarlo committed:
Pep8 corrections

Commit:
Merging

Commit:
Merging

Commit:
Add recursive maximal size to True

Commit:
Creation AdaptiveRefinement class to perform adaptive refinement

Commit:
Add store_lower_levels_samples flag

riccardotosi committed:
Default MC finer mesh and change confidence MLMC

Commit:
Add KratosMultiphysics

AlejandroCornejo commented:
Since the changes in the core have been removed in this PR, the @KratosMultiphysics/implementation-committee approves the changes (@adityaghantasala has already expressed his agreement based on the this change).  Waiting for the final ok from the @KratosMultiphysics/technical-committee

riccardotosi commented:
@RiccardoRossi ping

Commit:
Set finer mesh as default.
Change tol0 and tolF to more appropriate values.
Correct bug on convergence computation, not we check convergence with tolF and not with tol_iteration.

Commit:
Update final tolerance

roigcarlo reviewed (approved):
All good from the @KratosMultiphysics/technical-committee. Sorry for taking that long to review

RiccardoRossi commented:
+1!!!
